src/v/cluster/data_migration_table.h

```diff
@@ -0,0 +1,135 @@
+
+    static constexpr auto commands = make_commands_list<
+      create_data_migration_cmd,
+      update_data_migration_state_cmd,
+      remove_data_migration_cmd>();
+
+    bool is_batch_applicable(const model::record_batch& b) const {
+        return b.header().type == model::record_batch_type::data_migration_cmd;
+    }
+    // an entry point for controller stm, receives a record batch if is
+    // applicable to data migration state machine.
+    ss::future<std::error_code> apply_update(model::record_batch);
+
+    /**
+     * Fills the snapshot with data from the migration table when requested to
+     * do so by the controller state machine
+     */
+    ss::future<> fill_snapshot(controller_snapshot&) const;
+    /**
+     * Called withe controller snapshot when the controller state machine is
+     * replied.
+     */
+    ss::future<> apply_snapshot(model::offset, const controller_snapshot&);
+    /**
+     * Returns a single migration with requested id or empty optional if no
+     * migration is found.
+     */
+    std::optional<std::reference_wrapper<const data_migration_metadata>>
+      get_migration(data_migration_id) const;
+
+    /**
+     * Registers notification that if fired every time a data migration state is
+     * updated.
+     *
+     * @return id of notification that can be used to unregister the
+     * notification
+     */
+    notification_id register_notification(notification_callback);
+
+    /**
+     * Unregisters notification with requested id.
+     */
+    void unregister_notification(notification_id);
+
+    /**
+     * Returns metadata of all currently active data migrations.
+     */
+    chunked_vector<data_migration_metadata> list_migrations() const {
+        chunked_vector<data_migration_metadata> ret;
+        ret.reserve(_migrations.size());
+        for (const auto& [_, meta] : _migrations) {
+            ret.push_back(meta.copy());
+        }
+        return ret;
+    }
+
+private:
+    friend class data_migration_frontend;
+    friend testing_details::data_migration_table_test_accessor;
+
+    ss::future<std::error_code> apply(create_data_migration_cmd);
+    ss::future<std::error_code> apply(update_data_migration_state_cmd);
+    ss::future<std::error_code> apply(remove_data_migration_cmd);
+
+    data_migration_id get_next_id() { return _next_id++; }
+    void notify_callbacks(data_migration_id id);
+
+    std::optional<validation_error>
+    validate_migrated_resources(const inbound_data_migration&) const;
+    std::optional<validation_error>
+    validate_migrated_resources(const outbound_data_migration&) const;
+
+    static bool is_valid_state_transition(
+      data_migration_state current_state, data_migration_state target_state);
+
+    static bool is_empty_migration(const data_migration&);
+
+    std::optional<validation_error>
+    validate_migrated_resources(const data_migration&) const;
+
+private:
+    data_migration_id _next_id{0};
+    data_migration_id _last_applied{};
```

Comment:
what is the expected difference here for initialization values?

Reply:
_last_applied is intialized to min value of int64_t to indicate that nothing was applied, while we start assinging migration ids from 0